### ISSUE TITLE
Stops playing after you repeat playing from resource file

### ISSUE THREAD

**Issue by iot-playground** [adrianstevens/Xamarin-Plugins#4]
Thank you for great plugin. I found some issues after repeating playing from resource on Android device - cache files are not cleaned and player stops playing after about 20 repetitions. This code works for me - my fixes are under JAI comment.

```
using Android.Content.Res;
using Plugin.SimpleAudioPlayer.Abstractions;
using System;
using System.IO;

namespace Plugin.SimpleAudioPlayer
{
    /// <summary>
    /// Implementation for Feature
    /// </summary>
    public class SimpleAudioPlayerImplementation : ISimpleAudioPlayer
    {
        ///<Summary>
        /// Raised when audio playback completes successfully 
        ///</Summary>
        public event EventHandler PlaybackEnded;

        Android.Media.MediaPlayer player;

        static int index = 0;

        //JAI
        System.Collections.Generic.Dictionary<int, string> cacheFiles = new System.Collections.Generic.Dictionary<int, string>();

        ///<Summary>
        /// Length of audio in seconds
        ///</Summary>
        public double Duration
        { get { return player == null ? 0 : ((double)player.Duration) / 1000.0; } }

        ///<Summary>
        /// Current position of audio playback in seconds
        ///</Summary>
        public double CurrentPosition
        { get { return player == null ? 0 : ((double)player.CurrentPosition) / 1000.0; } }

        ///<Summary>
        /// Playback volume (0 to 1)
        ///</Summary>
        public double Volume
        {
            get { return _volume; }
            set { SetVolume(_volume = value, Balance); }
        }
        double _volume = 0.5;

        ///<Summary>
        /// Balance left/right: -1 is 100% left : 0% right, 1 is 100% right : 0% left, 0 is equal volume left/right
        ///</Summary>
        public double Balance
        {
            get { return _balance; }
            set { SetVolume(Volume, _balance = value); }
        }
        double _balance = 0;

        ///<Summary>
        /// Indicates if the currently loaded audio file is playing
        ///</Summary>
        public bool IsPlaying
        { get { return player == null ? false : player.IsPlaying; } }

        public bool Loop
        {
            get { return _loop; }
            set { _loop = value;  if (player != null) player.Looping = _loop; }
        }
        bool _loop;

        ///<Summary>
        /// Indicates if the position of the loaded audio file can be updated
        ///</Summary>
        public bool CanSeek
        { get { return player == null ? false : true; } }

        string path;

        ///<Summary>
        /// Load wav or mp3 audio file as a stream
        ///</Summary>
        public bool Load(Stream audioStream)
        {
            //cache to the file system
            path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.Personal), $"cache{index++}.wav");
            var fileStream = File.Create(path);
            audioStream.CopyTo(fileStream);
            fileStream.Close();
            //JAI
            audioStream.Close();

            //load the cached audio into MediaPlayer
            player?.Dispose();
            player = new Android.Media.MediaPlayer() { Looping = Loop };
            player?.SetDataSource(path);
            player?.Prepare();

            if (player != null)
            {
                player.Completion += OnPlaybackEnded;
                //JAI
                cacheFiles.Add(player.AudioSessionId, path);
            }
            return (player == null) ? false : true;
        }

        ///<Summary>
        /// Load wav or mp3 audio file from the iOS Resources folder
        ///</Summary>
        public bool Load(string fileName)
        {
            AssetFileDescriptor afd = Android.App.Application.Context.Assets.OpenFd(fileName);

            player?.Dispose();
            player = new Android.Media.MediaPlayer() { Looping = Loop };

            player?.SetDataSource(afd.FileDescriptor, afd.StartOffset, afd.Length);
            
            player?.Prepare();

            if (player != null)
                player.Completion += OnPlaybackEnded;

            return (player == null) ? false : true;
        }

        ///<Summary>
        /// Begin playback or resume if paused
        ///</Summary>
        public void Play()
        {
            if (player == null)
                return;

            if (player.IsPlaying)
            {
                player.Pause();
                player.SeekTo(0);
            }

            player.Start();
        }

        ///<Summary>
        /// Stop playack and set the current position to the beginning
        ///</Summary>
        public void Stop()
        {
            player?.Pause();
            player?.SeekTo(0);
        }

        ///<Summary>
        /// Pause playback if playing (does not resume)
        ///</Summary>
        public void Pause()
        {
            player?.Pause();
        }

        ///<Summary>
        /// Set the current playback position (in seconds)
        ///</Summary>
        public void Seek (double position)
        {
            if (player != null)
                player.SeekTo((int)position*1000);
        }

        ///<Summary>
        /// Sets the playback volume as a double between 0 and 1
        /// Sets both left and right channels
        ///</Summary>
        void SetVolume(double volume, double balance)
        {
            volume = Math.Max(0, volume);
            volume = Math.Min(1, volume);

            balance = Math.Max(0, balance);
            balance = Math.Min(1, balance);

            var right = (balance < 0) ? volume * -1 * balance : volume;
            var left = (balance > 0) ? volume * 1 * balance : volume;

            player.SetVolume((float)left, (float)right);
        }

        void OnPlaybackEnded(object sender, EventArgs e)
        {         
            PlaybackEnded?.Invoke(sender, e);
            // JAI release cache file and player
            var pl = (Android.Media.MediaPlayer)sender;

            if (cacheFiles.ContainsKey(pl.AudioSessionId))
            {
                var path = cacheFiles[pl.AudioSessionId];
                cacheFiles.Remove(pl.AudioSessionId);
                if (File.Exists(path))
                    File.Delete(path);
            }

            pl.Release();
            pl.Dispose();
        }


        bool isDisposed = false;

        protected virtual void Dispose(bool disposing)
        {
            if (isDisposed || player == null)
                return;

            if (disposing)
            {
                if (IsPlaying)
                    player.Stop();

                player.Completion -= OnPlaybackEnded;

                player.Dispose();
            }
            player = null;

            isDisposed = true;

            if (string.IsNullOrWhiteSpace(path) == false)
            {
                try
                {
                    File.Delete(path);
                    path = string.Empty;
                }
                catch
                {
                }
            }

            isDisposed = true;
        }

        ~SimpleAudioPlayerImplementation()
        {
            Dispose(false);
        }

        public void Dispose()
        {
            Dispose(true);

            GC.SuppressFinalize(this);
        }
    }
}
```

**Comment by odahan** [comment]
Hi,
If you release the cache and the player in OnPlaybackEnded method as you do here, can you reuse the player with a Play() or the later will crash ? As I see your code you must Load again to play the same thing.
Am I missing something ?
Thanks in advance !

**Comment by adrianstevens** [comment]
@iot-playground fantastic - thank you for the code

I like the strategy - as Odahan mentioned, the only downside is the needing to reloading the audio source after each play. I believe I have an alternative working fix (see issue #6) which I'll publish soon.

Let's leave this issue open until it's been tested and confirmed. And if needed I'll merge in your code.

Re closing the Stream (audioStream). I debated this when I wrote the code originally but I decided against it; Imo the calling code "owns" the stream and should be responsible for cleanup (i.e. your shared code). I'd love to hear arguments if others feel differently.
